pkg/lib/repo/repository.go

```diff
@@ -0,0 +1,270 @@
+	// If the entire blob is going to be less than 100 MiB, upload it in one piece. Otherwise,
+	if expected.Size < 100<<20 {
+		return r.uploadBlobMonolithic(ctx, location, postResp, expected, content)
+	} else {
+		return r.uploadBlobChunked(ctx, location, postResp, expected, content)
+	}
+}
+
+// uploadBlobMonolithic performs a monolithic blob upload as per the distribution spec. The content of the blob is uploaded
+// in one PUT request at the provided location.
+func (r *Repository) uploadBlobMonolithic(ctx context.Context, location *url.URL, postResp *http.Response, expected ocispec.Descriptor, content io.Reader) (string, error) {
+	req, err := http.NewRequestWithContext(ctx, http.MethodPut, location.String(), content)
+	if err != nil {
+		return "", err
+	}
+	// Set Content-Length header
+	if req.GetBody != nil && req.ContentLength != expected.Size {
+		// short circuit a size mismatch for built-in types.
+		return "", fmt.Errorf("mismatch content length %d: expect %d", req.ContentLength, expected.Size)
+	}
+	req.ContentLength = expected.Size
+
+	// Set Content-Type to required 'application/octet-stream'
+	req.Header.Set("Content-Type", "application/octet-stream")
+
+	// Set digest query to mark this as completing the upload
+	q := req.URL.Query()
+	q.Set("digest", expected.Digest.String())
+	req.URL.RawQuery = q.Encode()
+
+	// Reuse credentials from POST request that initiated upload
+	if auth := postResp.Request.Header.Get("Authorization"); auth != "" {
+		req.Header.Set("Authorization", auth)
+	}
+
+	output.SafeDebugf("Uploading blob as one chunk")
+	// TODO: Handle warnings from remote
+	// References:
+	//   - https://github.com/opencontainers/distribution-spec/blob/v1.1.0-rc4/spec.md#warnings
+	//   - https://www.rfc-editor.org/rfc/rfc7234#section-5.5
+	resp, err := r.client().Do(req)
+	if err != nil {
+		return "", fmt.Errorf("failed to upload blob: %w", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusCreated {
+		return "", handleRemoteError(resp)
+	}
+
+	blobLocation, err := resp.Location()
+	if err != nil {
+		output.Errorf("Warning: remote registry did not return blob location")
+	}
+
+	return blobLocation.String(), nil
+}
+
+// uploadBlobChunked performs a chunked blob upload as per the distribution spec. The blob is divided into chunks of maximum 100MiB
+// in size and uploaded sequentially through PATCH requests. Once entire blob is uploaded, a PUT request marks the upload as complete.
+// Note that the distribution spec 1) requires blobs to uploaded in-order, and 2) does not have a way of specifying maximum blob
+// size.
+func (r *Repository) uploadBlobChunked(ctx context.Context, location *url.URL, postResp *http.Response, expected ocispec.Descriptor, content io.Reader) (string, error) {
+	// TODO: Handle 'OCI-Chunk-Min-Length' header in post response
```

Comment:
We can not leave this as todo, ghcr push for instance fails for me with `failed to copy to remote: requested_range_not_satisfiable: the request body is too large and exceeds the maximum permissible limit of 4.00MiB`

Reply:
Hm, that's a little strange, since this TODO covers _minimum_ length. The OCI spec has no way to specify _maximum_ length.

Does ghcr really limit chunks to 4.00 MiB?

Reply:
That is right... this is the min.

Reply:
Checked again it gets a 416 from ghcr with the detail above.

Reply:
Checking the response headers, ghcr.io says nothing to tell us a maximum size :upside_down_face: 

I'm going to special-case it for now.

Reply:
Added https://github.com/jozu-ai/kitops/pull/329/commits/46f3c7ccb6cbf37f98b5d21050e49f23022e16dd to let us override per-repo, as this seems to be the only way to do this. This effectively reverts to the old way for ghcr.io only.

Checked with podman, and they use `transfer-encoding: chunked` + patch to upload large layers in one go, so we can implement that in the future and potentially use it instead.